Test afterSubscriptionIsCancelledRequestMustBeNops failed

The infamous bug has cropped up again 😞 :
```
[info] - required_spec306_afterSubscriptionIsCancelledRequestMustBeNops *** FAILED ***
[info]   java.lang.AssertionError: Did not expect an element but got element [1]
```
This occurred when running locally on the scala 2.11 build.

I thought a solution was found in #29, but it looks like more investigation is needed.

--- a/core/src/main/scala/fs2/interop/reactivestreams/StreamSubscription.scala
+++ b/core/src/main/scala/fs2/interop/reactivestreams/StreamSubscription.scala
@@ -24,16 +24,17 @@ final class StreamSubscription[F[_], A](
     extends Subscription {
   import StreamSubscription._
 
-  def onError(e: Throwable) = F.delay(sub.onError(e)) *> cancelled.set(true)
+  // We want to make sure `cancelled` is set _before_ signalling the subscriber
+  def onError(e: Throwable) = cancelled.set(true) *> F.delay(sub.onError(e))
+  def onComplete = cancelled.set(true) *> F.delay(sub.onComplete)
 
   def unsafeStart(): Unit = {
     def subscriptionPipe: Pipe[F, A, A] =
       in => {
         def go(s: Stream[F, A]): Pull[F, A, Unit] =
           Pull.eval(requests.dequeue1).flatMap {
-            case Cancelled => Pull.done
-            case InfiniteRequests => s.pull.echo
-            case FiniteRequests(n) =>
+            case Infinite => s.pull.echo
+            case Finite(n) =>
               s.pull.take(n).flatMap {
                 case None => Pull.done
                 case Some(rem) => go(rem)
@@ -49,25 +50,32 @@ final class StreamSubscription[F[_], A](
         .interruptWhen(cancelled)
         .evalMap(x => F.delay(sub.onNext(x)))
         .handleErrorWith(e => Stream.eval(onError(e)))
-        .onFinalize {
-          cancelled.get.ifM(
-            ifTrue = F.unit,
-            ifFalse = cancelled.set(true) *> F.delay(sub.onComplete)
-          )
-        }
+        .onFinalize(cancelled.get.ifM(ifTrue = F.unit, ifFalse = onComplete))
         .compile
         .drain
 
     async.unsafeRunAsync(s)(_ => IO.unit)
   }
 
+  // According to the spec, it's acceptable for a concurrent cancel to not
+  // be processed immediately, bt if you have synchronous `cancel();
+  // request()`, then the request _must_ be a no op. For this reason, we
+  // need to make sure that `cancel()` does not return until the
+  // `cancelled` signal has been set.
+  // See https://github.com/zainab-ali/fs2-reactive-streams/issues/29
+  // and https://github.com/zainab-ali/fs2-reactive-streams/issues/46
   def cancel(): Unit =
-    async.unsafeRunAsync(cancelled.set(true) *> requests.enqueue1(Cancelled))(_ => IO.unit)
+    IO.async[Unit] { cb =>
+        async.unsafeRunAsync {
+          cancelled.set(true) *> F.delay(cb(Right()))
+        }(_ => IO.unit)
+      }
+      .unsafeRunSync
 
   def request(n: Long): Unit = {
     val request =
-      if (n == java.lang.Long.MAX_VALUE) InfiniteRequests.pure[F]
-      else if (n > 0) FiniteRequests(n).pure[F]
+      if (n == java.lang.Long.MAX_VALUE) Infinite.pure[F]
+      else if (n > 0) Finite(n).pure[F]
       else F.raiseError(new IllegalArgumentException(s"3.9 - invalid number of elements [$n]"))
 
     val prog = cancelled.get
@@ -79,20 +87,10 @@ final class StreamSubscription[F[_], A](
 
 object StreamSubscription {
 
-  /** Represents a request to publish elements by a downstream subscriber */
+  /** Represents a downstream subscriber's request to publish elements */
   sealed trait Request
-
-  /** The downstream reactivestreams subscriber has requested an infinite number of elements */
-  case object InfiniteRequests extends Request
-
-  /** The downstream subscriber has requested a finite number of elements.
-    *
-    * @param n the number of elements requested
-    */
-  case class FiniteRequests(n: Long) extends Request
-
-  /** The downstream subscriber has cancelled the subscription. */
-  case object Cancelled extends Request
+  case object Infinite extends Request
+  case class Finite(n: Long) extends Request
 
   def apply[F[_]: Effect, A](sub: Subscriber[A], stream: Stream[F, A])(
     implicit ec: ExecutionContext

--- a/core/src/test/scala/fs2/interop/reactivestreams/CancellationSpec.scala
+++ b/core/src/test/scala/fs2/interop/reactivestreams/CancellationSpec.scala
@@ -1,0 +1,58 @@
+package fs2
+package interop
+package reactivestreams
+
+import org.reactivestreams._
+import cats.effect._
+
+import scala.concurrent.ExecutionContext.Implicits.global
+import java.util.concurrent.atomic.AtomicBoolean
+
+import org.scalatest.FunSuite
+
+/**
+  * This behaviour is already tested by the Reactive Stream test
+  * suite, but it's proven difficult to enforce, so we add our own
+  * tests that run the assertions multiple times to make possible
+  * failures due to race conditions more repeatable
+  */
+class CancellationSpec extends FunSuite {
+  case class Sub[A](b: AtomicBoolean) extends Subscriber[A] {
+    def onNext(t: A) = b.set(true)
+    def onComplete() = b.set(true)
+    def onError(e: Throwable) = b.set(true)
+    def onSubscribe(s: Subscription) = b.set(true)
+  }
+
+  val s = Stream.range(0, 5).covary[IO]
+
+  val attempts = 10000
+
+  test("after subscription is cancelled request must be noOps") {
+    var i = 0
+    val b = new AtomicBoolean(false)
+    while(i < attempts) {
+      val sub = StreamSubscription(Sub[Int](b), s).unsafeRunSync
+      sub.unsafeStart
+      sub.cancel
+      sub.request(1)
+      sub.request(1)
+      sub.request(1)
+      i = i + 1
+    }
+    if (b.get) fail("onNext was called after the subscription was cancelled")
+  }
+
+  test("after subscription is cancelled additional cancelations must be noOps") {
+    var i = 0
+    val b = new AtomicBoolean(false)
+    while(i < attempts) {
+      val sub = StreamSubscription(Sub[Int](b), s).unsafeRunSync
+      sub.unsafeStart
+      sub.cancel
+      sub.cancel
+      i = i + 1
+    }
+    if (b.get) fail("onCancel was called after the subscription was cancelled")
+  }
+}

